satellite: graceful evacuation when using ClusterAPI

WanzenBug opened the pull request:
In a ClusterAPI-managed cluster, we want to support graceful evacuation of nodes. In particular, we want to ensure a smooth transfer of resources when new machines are rolled out and old ones are shut down.

To do that, we adapt the existing evacuation of satellites to better handle resource migration. The new evacuation performs 8 steps:

1. We ensure that existing PVs are reschedulable to other nodes. This is necessary for volumes using the "allowRemoteVolumeAccess: false" parameter. Otherwise, these PVs remain stuck to the node that is going to be removed.
2. We wait for all LINSTOR Satellite resources to report a ready state. This ensures that during a rolling restart/update of a cluster, the replacement node is already ready from LINSTOR side.
3. Allow ClusterAPI to drain the machine, moving workloads to replacement nodes.
4. Wait for all PVs that were previously attached on the evacuating Node to be attached on other Nodes. This helps LINSTOR with the following node evacuation.
5. Signal LINSTOR to start evacuating the Satellite. In particular, since the previously attached PVs are now used on a different node, LINSTOR will try to place the new replica on that particular node, if possible.
6. Wait for evacuationt to complete.
7. Remove any temporary annotations used to mark PVs are reschedulable. The "allowRemoteVolumeAccess: false" resources should have new diskfuls now, letting the Affinity Controller re-apply the strict node affinity.
8. Allow ClusterAPI to remove the node, also removing the Satellite.

WanzenBug committed:
satellite: graceful evacuation when using ClusterAPI

In a ClusterAPI-managed cluster, we want to support graceful evacuation of
nodes. In particular, we want to ensure a smooth transfer of resources when
new machines are rolled out and old ones are shut down.

To do that, we adapt the existing evacuation of satellites to better handle
resource migration. The new evacuation performs 8 steps:

1. We ensure that existing PVs are reschedulable to other nodes. This is
   necessary for volumes using the "allowRemoteVolumeAccess: false" parameter.
   Otherwise, these PVs remain stuck to the node that is going to be removed.
2. We wait for all LINSTOR Satellite resources to report a ready state. This
   ensures that during a rolling restart/update of a cluster, the replacement
   node is already ready from LINSTOR side.
3. Allow ClusterAPI to drain the machine, moving workloads to replacement
   nodes.
4. Wait for all PVs that were previously attached on the evacuating Node to
   be attached on other Nodes. This helps LINSTOR with the following node
   evacuation.
5. Signal LINSTOR to start evacuating the Satellite. In particular, since
   the previously attached PVs are now used on a different node, LINSTOR will
   try to place the new replica on that particular node, if possible.
6. Wait for evacuationt to complete.
7. Remove any temporary annotations used to mark PVs are reschedulable. The
   "allowRemoteVolumeAccess: false" resources should have new diskfuls now,
   letting the Affinity Controller re-apply the strict node affinity.
8. Allow ClusterAPI to remove the node, also removing the Satellite.

Co-authored-by: Joel Colledge <joel.colledge@linbit.com>
Signed-off-by: Moritz Wanzenböck <moritz.wanzenboeck@linbit.com>

WanzenBug commented:
@JoelColledge Sorry for the large diff, but I don't really see how this could be made in smaller increments. I'm not really happy with the duplication/general flow of the ClusterAPI/Machine stuff within the "reconcileSatelliteState" method, but I also don't see how it could be anywhere else. There is also the duplication with "deleteSatellite", but again, I don't have a clue on how to do better. 

I'm working on a explanatory text for how deletion policies and `deletionPolicy: Evacuate` work in particular.

JoelColledge commented on cmd/main.go:
```diff
@@ -104,9 +108,50 @@ func main() {
+
+		machineClientCfg, err = clientCfg.ClientConfig()
+		if err != nil {
+			setupLog.Error(err, "unable to parse cluster api kubeconfig")
```
This error message is used twice. Make them distinct to help identify the cause of errors?

JoelColledge commented on go.mod:
```diff
@@ -52,9 +55,13 @@ require (
 	github.com/google/go-querystring v1.1.0 // indirect
 	github.com/google/pprof v0.0.0-20250403155104-27863c87afa6 // indirect
 	github.com/google/uuid v1.6.0 // indirect
+	github.com/haySwim/data v0.2.0 // indirect
```
I wonder how this indirect dependency got in? A module from Hayley, untouched for over 6 years :smile:

Just to be clear: I don't see any cause for concern with the dependency itself.

WanzenBug replied:
I think that is because we now use some additional packages from the `linstor-csi` project. We use that project there, I guess you could consider that project complete :smile:

JoelColledge commented on internal/controller/linstorsatellite_controller.go:
```diff
@@ -512,9 +558,14 @@ func (r *LinstorSatelliteReconciler) reconcileStoragePools(ctx context.Context,
 }
 
-func (r *LinstorSatelliteReconciler) deleteSatellite(ctx context.Context, lsatellite *piraeusiov1.LinstorSatellite) error {
+func (r *LinstorSatelliteReconciler) deleteSatellite(ctx context.Context, lsatellite *piraeusiov1.LinstorSatellite, node *corev1.Node, conds conditions.Conditions) (string, bool, error) {
```
Document the return values?

JoelColledge commented on pkg/evacuation/evacuation.go:
```diff
@@ -0,0 +1,364 @@
+
+		if pv.Annotations == nil {
+			pv.Annotations = make(map[string]string)
+			toUpdate = true
```
Is converting a nil value into an empty map a reason to update? When we update with an empty map and then `Get` again, do we get an empty map back?

WanzenBug replied:
It is not. It would generally cause an update, as with an empty map we are guaranteed to also trigger the following updates, but I removed this here still.

JoelColledge commented on internal/controller/linstorsatellite_controller.go:
```diff
@@ -311,6 +317,32 @@ func (r *LinstorSatelliteReconciler) kustomizeNodeResources(ctx context.Context,
 }
 
 func (r *LinstorSatelliteReconciler) reconcileLinstorSatelliteState(ctx context.Context, lsatellite *piraeusiov1.LinstorSatellite, node *corev1.Node, conds conditions.Conditions) error {
+	machine, err := r.MachineClient.GetMachineForNode(ctx, node)
```
Perhaps a brief comment explaining why it is OK to call this regardless of whether ClusterAPI integration is configured?

JoelColledge commented on pkg/evacuation/evacuation.go:
```diff
@@ -0,0 +1,364 @@
+}
+
+func evacuateNode(ctx context.Context, lclient *lapi.Client, node *lapi.Node) error {
+	// We retry evacuation every time. LINSTOR needs multiple kicks until it actually starts moving all resources in
+	// some cases.
```
Perhaps this needs fixing in LINSTOR :joy:

JoelColledge commented on pkg/evacuation/evacuation.go:
```diff
@@ -0,0 +1,364 @@
+
+func getAttachedNodes(pvName string, attachments []storagev1.VolumeAttachment) []string {
+	var attachedNodes []string
+	for _, attachment := range attachments {
```
It looks like we iterate over all the attachments in the cluster for all the PVs on the node. It seems realistic to me that we could have 100s of PVs on a node and 1000s of attachments in total. That makes something on the order of millions of iterations of the inner loop each time we reconcile while in this state. Right? Should we do it more efficiently? I'm sure it is possible with one iteration over the attachments, building a map by PV or something.

WanzenBug replied:
> I'm sure it is possible with one iteration over the attachments, building a map by PV or something.

Done just that :+1:

JoelColledge commented on pkg/evacuation/evacuation_test.go:
```diff
@@ -0,0 +1,434 @@
+		unattachedRemotePV,
+	)
+
+	// Create fake linstor resources, let the "unattached" ones be still on node A.
```
Update copied comment.

JoelColledge reviewed (commented):
I have looked through this and it seems pretty clean to me :+1:

WanzenBug committed:
docs: add explanation for deletion policies

Co-authored-by: Joel Colledge <joel.colledge@linbit.com>
Signed-off-by: Moritz Wanzenböck <moritz.wanzenboeck@linbit.com>

WanzenBug committed:
webhook: fix a crash when validating change pool types

When a user updated the storage pool to a different type, we did not properly
protect deep-comparisons of old configuration against the new configuration.
This meant that the webhook would crash when trying to update the storage pool
type, instead of reporting a human-readable error.

Fix that by guarding before making deep storage pool type comparisons and
adding a test to exercise that code path.

Signed-off-by: Moritz Wanzenböck <moritz.wanzenboeck@linbit.com>

WanzenBug commented:
@JoelColledge please re-review

I added docs around the deletion process. I also fixed a super small issue in our webhooks that are not actually related to this feature, just noticed it during testing...

JoelColledge commented on docs/explanation/deletion-policies.md:
```diff
@@ -0,0 +1,117 @@
+1. Piraeus Datastore searches for PersistentVolumes (PVs) that:
+    * Are attached on the Node to be evacuated and marks them with `wait-for-reattach.evacuation.piraeus.io/<node-name>`
+      annotation. It will later use this annotation to wait for the PV to be reattached.
+    * In addition, if the PVS are also either using the `allowRemoteVolumeAcesss: "false"` policy or are otherwise bound
```
```suggestion
    * In addition, if the PVs are also either using the `allowRemoteVolumeAcesss: "false"` policy or are otherwise bound
```

JoelColledge commented on docs/explanation/deletion-policies.md:
```diff
@@ -0,0 +1,117 @@
+      to the Node to be evacuated in their `.spec.nodeAffinity`, they are annotated with
+      `override.piraeus.io/<nodename>`. This causes the LINSTOR Affinity Controller to temporarily mark the PV as
+      accessible from any Node.
+2. Piraeus Datastore will wait for any other `LinstorSatellite` resource to become available. This ensures that during
```
```suggestion
2. Piraeus Datastore will wait for all other `LinstorSatellite` resources to become available. This ensures that during
```

Right?

JoelColledge commented on docs/explanation/deletion-policies.md:
```diff
@@ -0,0 +1,117 @@
+   and diskless resources on the node. Volumes that are currently in use on a diskless will get a local replica, if that
+   is possible based on the available storage pools.
+6. Piraeus Datastore will wait for LINSTOR to complete the evacuation. The evacuated Node no longer has any resources
+   or snapshot.
```
```suggestion
   or snapshots.
```

JoelColledge commented on docs/explanation/deletion-policies.md:
```diff
@@ -0,0 +1,117 @@
+!!! warning "There are no checks preventing deletion of the last replica of a volume"
+
+    Using this policy means that there are no checks the prevent the deletion of the last replica of a volume. While
+    LINSTOR automatically tries to maintain the requested number of replicas, does so only periodically.
```
```suggestion
    LINSTOR automatically tries to maintain the requested number of replicas, it does so only periodically.
```

JoelColledge commented on pkg/evacuation/evacuation.go:
```diff
@@ -0,0 +1,367 @@
+// * marks them with annotations so later steps can find them.
+// * ensures that "local" PVs are temporarily reschedulable during evacuation.
+func preparePVsforEvacuation(ctx context.Context, cl client.Client, nodeName string) (string, bool, error) {
+	var volumeAttachementList storagev1.VolumeAttachmentList
```
```suggestion
	var volumeAttachmentList storagev1.VolumeAttachmentList
```

JoelColledge commented on pkg/evacuation/evacuation.go:
```diff
@@ -0,0 +1,367 @@
+		return "", false, err
+	}
+
+	var volumeAttachementList storagev1.VolumeAttachmentList
```
```suggestion
	var volumeAttachmentList storagev1.VolumeAttachmentList
```

JoelColledge reviewed (commented):
The new explanation doc is great :100:

Just noticed a few tiny details.

WanzenBug commented:
Thanks for the attentive review! I squashed the suggested changes into the right commits.
